apps/rush-lib/src/logic/pnpm/PnpmProjectDependencyManifest.ts

```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * If the project/.rush/temp/project-dependencies.json file exists, delete it. Otherwise, do nothing.
+   */
+  public deleteIfExists(): void {
+    FileSystem.deleteFile(this._projectDependencyManifestFilename, { throwIfNotExists: false });
+  }
+
+  private _addDependencyInternal(
+    name: string,
+    version: string,
+    parentShrinkwrapEntry: IPnpmShrinkwrapDependencyYaml,
+    throwIfShrinkwrapEntryMissing: boolean = true
+  ): void {
+    const specifier: string = `${name}@${version}`;
+    if (!PnpmProjectDependencyManifest._cache.has(specifier)) {
+      const shrinkwrapEntry: IPnpmShrinkwrapDependencyYaml | undefined = this._pnpmShrinkwrapFile.getShrinkwrapEntry(
+        name,
+        version
+      );
+
+      if (!shrinkwrapEntry) {
+        if (throwIfShrinkwrapEntryMissing) {
+          throw new InternalError(`Unable to find dependency ${name} with version ${version} in shrinkwrap.`);
+        }
+        return;
+      }
+
+      PnpmProjectDependencyManifest._cache.set(specifier, shrinkwrapEntry.resolution.integrity);
+
+      for (const dependencyName in shrinkwrapEntry.dependencies) {
+        if (shrinkwrapEntry.dependencies.hasOwnProperty(dependencyName)) {
+          const dependencyVersion: string = shrinkwrapEntry.dependencies[dependencyName];
+          this._addDependencyInternal(dependencyName, dependencyVersion, shrinkwrapEntry);
+        }
+      }
+
+      for (const optionalDependencyName in shrinkwrapEntry.optionalDependencies) {
+        if (shrinkwrapEntry.optionalDependencies.hasOwnProperty(optionalDependencyName)) {
+          // Optional dependencies may not exist. Don't blow up if it can't be found
+          const dependencyVersion: string = shrinkwrapEntry.optionalDependencies[optionalDependencyName];
+          this._addDependencyInternal(
+            optionalDependencyName,
+            dependencyVersion,
+            shrinkwrapEntry,
+            throwIfShrinkwrapEntryMissing = false);
+        }
+      }
+
+      for (const peerDependencyName in shrinkwrapEntry.peerDependencies) {
+        if (shrinkwrapEntry.peerDependencies.hasOwnProperty(peerDependencyName)) {
+          // Peer dependencies come in the form of a semantic version range
+          const dependencySemVer: string = shrinkwrapEntry.peerDependencies[peerDependencyName];
+          // Check the current package to see if the dependency is already satisfied
+          if (
+            shrinkwrapEntry.dependencies &&
+            shrinkwrapEntry.dependencies.hasOwnProperty(peerDependencyName)
+          ) {
+            const dependencySpecifier: DependencySpecifier | undefined = parsePnpmDependencyKey(
+              peerDependencyName,
+              shrinkwrapEntry.dependencies[peerDependencyName]
+            );
+            if (
+              dependencySpecifier &&
+              semver.valid(dependencySpecifier.versionSpecifier) &&
+              semver.satisfies(dependencySpecifier.versionSpecifier, dependencySemVer)
```

Comment:
I feel like we shouldn't ever need `semver.satisifes()`, since NodeJS module reasolution doesn't consider  that.  And when `strictPeerDependencies=false`, the shrinkwrap file may need to record versions that are inconsistent with package.json.  Throwing the `InternalError` would be a problem in that situation.

`semver.valid()` is probably an okay sanity check though, in case we encounter a file format case that we haven't seen before.